src/ConnectionManager.js

```diff
@@ -28,141 +24,162 @@ export default class ConnectionManager {
+        );
+        connection.addListener(
+          ConnectionEvent.COMPLETE,
+          context.handleConnectionClose
+        );
+        connection.addListener(
+          ConnectionEvent.ERROR,
+          context.handleConnectionClose
+        );
       },
 
       /**
-       * @property {function} handleConnectionError
-       * @name ConnectionManager~Context#handleConnectionError
-       * @param {ConnectionEvent} event
+       * Removes Listeners from Connection
+       * 
+       * @param {AbstractConnection} connection 
        */
-      handleConnectionError: event => {
-        this.dequeue(event.target);
-      }
-    };
-
-    this.enqueue = this.enqueue.bind(context);
-    this.dequeue = this.dequeue.bind(context);
-  }
+      removeListeners(connection) {
+        connection.removeListener(
+          ConnectionEvent.ABORT,
+          context.handleConnectionClose
+        );
+        connection.removeListener(
+          ConnectionEvent.COMPLETE,
+          context.handleConnectionClose
+        );
+        connection.removeListener(
+          ConnectionEvent.ERROR,
+          context.handleConnectionClose
+        );
+      },
 
-  /**
-   * Queues given connection with given priority
-   *
-   * @this ConnectionManager~Context
-   * @param {AbstractConnection} connection – connection to queue
-   * @param {Integer} [priority] – optional change of priority
-   */
-  enqueue(connection, priority) {
-    if (connection.state === AbstractConnection.CLOSED) {
-      return;
-    }
+      /**
+       * This function is where the magic happens.
+       */
+      loop() {
+        // store all open items
+        let openList = context.list.filter(listItem => 
+          listItem.connection.state === AbstractConnection.OPEN
+        );
+        
+        let waitingList = context.list
+          .filter(listItem => 
+            listItem.connection.state === AbstractConnection.INIT
+          )
+          // increase priority on old items
+          .map((listItem) => {
+            if (
+              listItem.priority < maxAutoPriority && 
+              listItem.created < Date.now() - maxAge
+            ) {
+              return new ConnectionQueueItem(listItem.connection, listItem.priority+1);
+            }
+            return listItem;
+          })
+          // and sort
+          .sortBy(listItem => -1 * listItem.priority);
+
+        // open a connection if possible
+        if (waitingList.size && openList.size < maxConnections) {
+          context.openConnection(waitingList.first().connection);
+        }
 
-    if (connection.state === AbstractConnection.INIT) {
-      this.waitingConnections = this.waitingConnections.enqueue(
-        connection,
-        priority
-      );
-    }
+        // merge them again.
+        context.list = openList.concat(waitingList);
+
+        // still waiting items? 
+        if(waitingList.size > 0) {
+          const delay = (maxConnections * openList.size);
```

Comment:
I tried to make this more "smart", but in the end it only increases complexity and doesnt realy brings a benefit…